src/tag_gc.rs

```diff
@@ -3,30 +3,120 @@ use rustc_data_structures::fx::FxHashSet;
 use crate::*;
 
 pub trait VisitMachineValues {
-    fn visit_machine_values(&self, visit: &mut impl FnMut(&Operand<Provenance>));
+    fn visit_machine_values(&self, visit: &mut ProvenanceVisitor);
+}
+
+pub trait MachineValue {
+    fn visit_provenance(&self, tags: &mut FxHashSet<SbTag>);
+}
+
+pub struct ProvenanceVisitor {
+    tags: FxHashSet<SbTag>,
+}
+
+impl ProvenanceVisitor {
+    pub fn visit<V>(&mut self, v: V)
+    where
+        V: MachineValue,
+    {
+        v.visit_provenance(&mut self.tags);
+    }
+}
+
+impl<T: MachineValue> MachineValue for &T {
+    fn visit_provenance(&self, tags: &mut FxHashSet<SbTag>) {
+        (**self).visit_provenance(tags);
+    }
+}
+
+impl MachineValue for Operand<Provenance> {
```

Comment:
Now that we have a bunch of stuff here, might be good to have a new file `machine_visitor.rs` or so.